.github/copilot-instructions.md

```diff
@@ -0,0 +1,837 @@
+
+// Receive with continuous callback pattern
+async_socket_receive_async(socket, receive_callback, context);
+```
+
+## THANDLE: Type-Safe Smart Pointers
+
+THANDLE provides C++-like smart pointer functionality for C, offering automatic memory management, type safety, and move semantics. It's a completely separate system from REFCOUNT and uses its own set of macros and functions.
+
+### Core THANDLE Concepts
+
+#### Declaration and Creation
+```c
+// Declare a THANDLE type for any structure
+typedef struct MY_OBJECT_TAG 
+{ 
+    int data; 
+    char* name;
+} MY_OBJECT;
+
+// Enable THANDLE support for the type (in header)
+THANDLE_TYPE_DECLARE(MY_OBJECT);
+
+// Define THANDLE implementation (in source file)
+THANDLE_TYPE_DEFINE(MY_OBJECT);
+
+// Create THANDLE instances using proper THANDLE macros
+THANDLE(MY_OBJECT) handle = THANDLE_MALLOC(MY_OBJECT)(dispose_function);  // Creates with ref count 1
+THANDLE(MY_OBJECT) null_handle = NULL;                                    // NULL handle
+
+// Create from existing content
+MY_OBJECT source = { .data = 42, .name = "test" };
+THANDLE(MY_OBJECT) copy_handle = THANDLE_CREATE_FROM_CONTENT(MY_OBJECT)(&source, dispose_function, copy_function);
+```
+
+### THANDLE Creation Methods
+
+#### THANDLE_MALLOC - Basic Object Creation
+```c
+// Simple object creation with dispose function
+THANDLE(MY_OBJECT) obj = THANDLE_MALLOC(MY_OBJECT)(dispose_function);
+
+// No dispose function needed (for simple data)
+THANDLE(MY_OBJECT) simple_obj = THANDLE_MALLOC(MY_OBJECT)(NULL);
+```
+
+#### THANDLE_MALLOC_FLEX - Flexible Array Members
+```c
+// For structs with flexible array members
+typedef struct STRING_OBJECT_TAG
+{
+    size_t length;
+    char data[];  // Flexible array member
+} STRING_OBJECT;
+
+// Allocate extra space for the flexible array
+THANDLE(STRING_OBJECT) str_obj = THANDLE_MALLOC_FLEX(STRING_OBJECT)(dispose_function, 100, sizeof(char));  // 100 chars
+```
+
+#### THANDLE_CREATE_FROM_CONTENT - Copy Construction
+```c
+// Copy from existing object
+MY_OBJECT original = { .data = 42, .name = "original" };
+THANDLE(MY_OBJECT) copy = THANDLE_CREATE_FROM_CONTENT(MY_OBJECT)(&original, dispose_function, copy_function);
+
+// Simple copy (memcpy) without custom copy function
+THANDLE(MY_OBJECT) simple_copy = THANDLE_CREATE_FROM_CONTENT(MY_OBJECT)(&original, dispose_function, NULL);
+```
+
+#### THANDLE_CREATE_FROM_CONTENT_FLEX - Flexible Array Copy
+```c
+// Copy flexible array structures with custom size calculation
+THANDLE(STRING_OBJECT) flex_copy = THANDLE_CREATE_FROM_CONTENT_FLEX(STRING_OBJECT)(
+    &original_string, 
+    dispose_function, 
+    copy_function, 
+    get_sizeof_function);  // Custom function to calculate total size
+```
+
+#### Assignment and Move Semantics
+```c
+// THANDLE_ASSIGN - Creates shared ownership (increments ref count)
+THANDLE(MY_OBJECT) handle1 = THANDLE_MALLOC(MY_OBJECT)(dispose_function);
+THANDLE(MY_OBJECT) handle2;
```

Comment:
probably compiler will already complain at this line...

Reply:
Added = NULL.